### PR TITLE
Add activation functions as fused operators to BatchNorm and Conv2d

### PR DESCRIPTION
**Pull request by wchao1115**
As part of this change, I generalize a way to allow an operator to be created before it is connected to the graph. This makes it possible for fused operators, particularly the various activation functions to be expressed by a generic operator type instead of through an ever-expanding list of enum values, which addresses both #185 and #138. 

Mechanically, a notion of `MLOperator` is defined. This is a definition of a type of operation and not the operation itself, as no input operands are needed in the creation of an operator and no operand is expected from it. The operator will then connect with the rest of the graph via the graph builder once the input operands are ready.  The best example of this mechanic is demonstrated in the note section of the `batchNormalization` opeation where an activation function is passed into as an `MLOperator` and not as `MLOperand`. The builder's method `setOperand` connects the operator with the input operands and get an output operand out of it. 

As for #138, much of the discussion in the issue is centered around the need to not prematurely leave out other possible activation functions that could be supported by the future recurrent network. But maintaining an ever-expanding list of enum values for all the possible (present and future) activation functions is error-prone and in a way giving out a wrong and confusing signal to the API implementer who certainly can only base their implementation on the current underlying platform support.

This change addresses the need to keep the list open while avoiding conveying an unrealistic expectation to the implementer by not defining the list of supported activation functions in the description of the operation itself but instead leaving it out to the caller. Note that the graph building operation would not fail even when an unsupported activation function is hooked up, but the subsequent graph compilation step `builder.build` would.

***
<a href="https://pr-preview.s3.amazonaws.com/webmachinelearning/webnn/pull/188.html" title="Last updated on Jul 11, 2021, 12:52 AM UTC (5bdc94b)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/webmachinelearning/webnn/188/a913875...5bdc94b.html" title="Last updated on Jul 11, 2021, 12:52 AM UTC (5bdc94b)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -1499,6 +1640,54 @@ partial interface MLGraphBuilder {
     tensor is specified by the *newShape* argument.
 </div>
 
+### setOperand ### {#api-mlgraphbuilder-set-operand}
```
Comment (huningxin) [inline index.bs:1643]: I suppose `setOperand` should be the implementation detail. 

It is because when building a graph, if the user code wants to fuse the activation, it would call the `MLOperator` version, e.g. `output = builder.conv2d(input, filter, {activation: builder.relu()})`. Otherwise, the caller would call the `MLOperand` version, e.g. `output = builder.relu(builder.conv2d(input, filter))`. The `setOperand` seems not useful for user code.

Any other usages I missed?

File: index.bs [hunk 2]
```diff
@@ -599,18 +607,18 @@ partial interface MLGraphBuilder {
     therefore its usage is encouraged from the performance standpoint.
     <pre highlight="js">
     const shape = [1,-1,1,1];
-    return builder.add(
+    return builder.setOperand(
+      activation,
```
Comment (huningxin) [inline index.bs:611]: As the sample code, I suppose it could just use one type of activation here, such as `builder.relu`.

File: index.bs [hunk 3]
```diff
@@ -529,6 +529,12 @@ dictionary MLOperandDescriptor {
 interface MLOperand {};
 </script>
 
+## MLOperator ## {#api-mloperator}
+<script type=idl>
+[SecureContext, Exposed=(Window, DedicatedWorker)]
+interface MLOperator {};
```
Comment (huningxin) [inline index.bs:535]: For consistency, should we name it as `MLOperation`? This spec uses "operation" term in other places.

File: index.bs (hunk 1 above)
Comment (wchao1115) [reply index.bs:1643]: Yes, in operations like `batchNorm` or `gruCell`. You're right it could be treated as a hidden concept for atomic operations like `conv2d`. For composite operations that happen to accept fused operators, it should be possible to implement them purely as composites of smaller ops, as the pseudo-code in their note sections suggests.

File: index.bs (hunk 3 above)
Comment (wchao1115) [reply index.bs:535]: Maybe it's just a wording thing but I tend to think of an *operation* as an *operator* with *operands* e.g. A + B is an operation, while the *operator* is just the act imposed on the operands e.g. + is the operator. 

I think of `conv2d` as an operation since it accepts operands as required params, while `relu` could either be an operation or an operator, depending on whether it is created with or without an operand. Does that make sense?

File: index.bs (hunk 2 above)
Comment (wchao1115) [reply index.bs:611]: The implementation of this operation will not know ahead of time what activation is parameterized here. This is why I think we also need to define `setOperand`.

File: index.bs (hunk 3 above)
Comment (huningxin) [reply index.bs:535]: Thanks for your explanation. It makes sense to me.

File: index.bs (hunk 2 above)
Comment (huningxin) [reply index.bs:611]: I suppose an implementation could know the activation type internally. For example, an implementation may add a internal slot into `MLOperator` that keeps the activation type, e.g. the `MLOperator` instance created by `builder.relu()` has the type `relu`. Later, when the implementation composites the `batchNormalization`, it can check the type of the `MLOperator`.

**Review by anssiko — APPROVED** [review]
Thank you for this substantial update!

I’ll unblock my review so you can go ahead after addressing all review feedback. Should this PR require more extensive discussion, we can bring this to our post-break call in mid-Aug.

File: index.bs (hunk 1 above)
Comment (gramalingam) [reply index.bs:1643]: How about a name like ```apply``` instead of ```setOperand```? I think it is better if we use a name that indicates the side-effect-free functional behavior rather than one that suggests an imperative behavior with side-effects. I think this is essentially the operation of applying a function (represented by operator) to its arguments.

Comment (wchao1115) [reply index.bs:1643]: Thanks for the feedback @gramalingam. I like that.

File: index.bs (hunk 2 above)
Comment (wchao1115) [reply index.bs:611]: What do you think about moving `setOperand` to the `MLOperator` itself instead of on the builder interface? That way the builder's implementation doesn't need to track the operator type of each operator internally, which could be tedious and ineffective. Instead just have the implementation of the supported operators (e.g. activation functions) support this additional capability themselves without effecting how other operations are implemented. This should address your concern, if I sense it correctly.

Comment (huningxin) [reply index.bs:611]: Moving `setOperand` to `MLOperator` looks better. However my point is either `MLOperator.setOperand/apply` or `MLOperator.type` should be internal to implementation. If there are no use cases for user code, I would suggest exclude it in this PR.

Comment (wchao1115) [reply index.bs:611]: An implementation that doesn't want to handle composite operations like `batchNorm` natively should be able to implement it as if it's a graph of smaller operations, like what being described in the Note section. Remember that that is one of our design principles -- that every big op defined must be publicly implementable by compositing smaller public constructs? 

Although we do expect operations like `batchNorm` to be implemented natively, some implementation on some platform may not have a mean to do that, and would only settle on a default implementation as a graph of public operations.

Comment (huningxin) [reply index.bs:611]: > An implementation that doesn't want to handle composite operations like `batchNorm` natively should be able to implement it as if it's a graph of smaller operations,  like what being described in the Note section.

How about the Note section says "When the input tensor is 4-D of the "nchw" layout and ___the activation is a relu operator___, the behavior of this operation can be generically emulated from the usage of other operations as follow."

```js
const shape = [1,-1,1,1];
return builder.relu(
  builder.add(
    builder.mul(
      builder.reshape(options.scale, shape),
      builder.div(
        builder.sub(input, builder.reshape(mean, shape)),
        builder.pow(
          builder.add(builder.reshape(variance, shape), builder.constant(options.epsilon)),
          builder.constant(0.5))
        )),
    builder.reshape(options.bias, shape)));
```

> some implementation on some platform may not have a mean to do that,

That's true. For example, the webnn-polyfill implementation uses TF.js's [`tf.batchNorm`](https://js.tensorflow.org/api/latest/#batchNorm) that doesn't support fused activation. 

> and would only settle on a default implementation as a graph of public operations

This might not be true. For example, webnn-polyfill implements WebNN spec defined ops as one or multiple TF.js kernels. For fused `batchNormalziation`, it could call the corresponding Tf.js kernel of an `MLOperator`, e.g. `tf.relu`, right after `tf.batchNorm`. That doesn't require the `setOperand` as a public API.

Comment (huningxin) [reply index.bs:611]: @wchao1115 , I put my idea into https://github.com/webmachinelearning/webnn/pull/193, please take a look. Thanks.

Comment (wchao1115) [reply index.bs]: Thanks for the feedback @huningxin . Can you please take a look at the new commit? I adapted your suggestion in this update.

**Review by huningxin — APPROVED** [review]
LGTM, thanks for the update!
